lib/worker/src/routes/vaults.rs

```diff
@@ -0,0 +1,460 @@
+                error: err.to_string(),
+            })
+        }
+    };
+
+    let records = match db::pub_cids(
+        &pool,
+        &vault,
+        params.limit(),
+        params.offset(),
+        params.before(),
+        params.after(),
+    )
+    .await
+    {
+        Ok(records) => records,
+        Err(_) => {
+            return HttpResponse::InternalServerError().json(ErrorResponse {
+                error: "failed to fetch deals".to_string(),
+            })
+        }
+    };
+
+    HttpResponse::Ok().json(records)
+}
+
+impl TryFrom<web::Path<std::string::String>> for Vault {
+    type Error = String;
+
+    fn try_from(value: web::Path<std::string::String>) -> Result<Vault, String> {
+        Vault::from(value.to_string())
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct FindRecordsByPubIdQueryParams {
+    limit: Option<i32>,
+    offset: Option<i32>,
+    before: Option<String>,
+    after: Option<String>,
+}
+
+impl FindRecordsByPubIdQueryParams {
+    fn limit(&self) -> i32 {
+        self.limit.unwrap_or(10)
+    }
+
+    fn offset(&self) -> i32 {
+        self.offset.unwrap_or(0)
+    }
+
+    fn before(&self) -> i64 {
+        if let Some(before) = &self.before {
+            let before = match before.parse::<i64>() {
+                Ok(timestamp) => return timestamp,
+                Err(_) => before,
+            };
+
+            if let Ok(before) = DateTime::parse_from_str(before.as_str(), "%Y-%m-%d") {
+                return before.timestamp();
+            }
+
+            if let Ok(before) = DateTime::parse_from_rfc3339(before.as_str()) {
+                return before.timestamp();
+            }
+
+            // if we could not parse the datetime, use the default
+            return 0;
+        }
+        0
+    }
+
+    fn after(&self) -> i64 {
+        if let Some(after) = &self.after {
+            let after = match after.parse::<i64>() {
+                Ok(timestamp) => return timestamp,
+                Err(_) => after,
+            };
+
+            if let Ok(after) = DateTime::parse_from_str(after.as_str(), "%Y-%m-%d") {
+                return after.timestamp();
+            }
+
+            if let Ok(after) = DateTime::parse_from_rfc3339(after.as_str()) {
+                return after.timestamp();
+            }
+
+            // if we could not parse the datetime, use the default
+            return 0;
+        }
+        0
+    }
+}
+
+pub async fn create_vault<E: EVMClient + 'static + std::marker::Sync>(
+    path: web::Path<String>,
+    input: web::Form<CreateVaultInput>,
+    evm_client: web::Data<E>,
+    pool: web::Data<PgPool>,
+) -> HttpResponse {
+    let vault: Vault = match path.try_into() {
+        Ok(p) => p,
+        Err(err) => {
+            return HttpResponse::BadRequest().json(ErrorResponse {
+                error: err.to_string(),
+            })
+        }
+    };
+
+    let account: String = match &input.account {
```

Comment:
i am thinking, instead of taking the address as an input maybe we can take the signature as an input and then recover address from the signature. this way this API will also authenticate the user.

Reply:
sounds right to me!

Reply:
what would they be signing? the post payload?

Reply:
yes, we need to think about some generic payload that could be verified. i think, maybe too much for this PR. so perhaps we can add it at some point later.

there are a few prior works that deal with this problem, we can burrow the payload structure from these 

- https://eips.ethereum.org/EIPS/eip-712
- https://eips.ethereum.org/EIPS/eip-191

Reply:
nice